util/conditions/patch.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conditions
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+)
+
+// PatchOperation defines patch operation types.
+type PatchOperation string
+
+const (
+	// AddConditionPatch defines an add condition patch operation.
+	AddConditionPatch PatchOperation = "Add"
+
+	// ChangeConditionPatch defines an change condition patch operation.
+	ChangeConditionPatch PatchOperation = "Change"
+
+	// RemoveConditionPatch defines a remove condition patch operation.
+	RemoveConditionPatch PatchOperation = "Remove"
+)
+
+// ConditionPatch define a patch that apply to a condition.
+type ConditionPatch struct {
+	Target *clusterv1.Condition
+	Base   *clusterv1.Condition
```

Comment:
Why do we need to store target/base?

Reply:
because I'm doing a three-way merge.
the merge patch will be built from the before object compared vs target, and the before conditions are the bases for condition patches.
then we will re-read conditions and by checking if something changed conditions in the meantime we can verify if the patch can merge or if there are merge confict

Reply:
Could we store this only once per patch set, instead of doing it for each patch?

Reply:
As in, a `Patch` should really become a list of operations